File: backend/scripts/check_coverage_ratchet.py
#!/usr/bin/env python3
import csv
import json
import sys
from pathlib import Path
from typing import Dict, Tuple


def load_baseline(path: Path) -> Dict[str, float]:
    with path.open("r", encoding="utf-8") as handle:
        payload = json.load(handle)
    return {"line": float(payload["line"]), "branch": float(payload["branch"])}


def load_current_coverage(csv_path: Path) -> Dict[str, float]:
    with csv_path.open("r", encoding="utf-8", newline="") as handle:
        rows = list(csv.DictReader(handle))
    line_missed = sum(int(row["LINE_MISSED"]) for row in rows)
    line_covered = sum(int(row["LINE_COVERED"]) for row in rows)
    branch_missed = sum(int(row["BRANCH_MISSED"]) for row in rows)
    branch_covered = sum(int(row["BRANCH_COVERED"]) for row in rows)

    line_total = line_missed + line_covered
    branch_total = branch_missed + branch_covered

    return {
        "line": 100.0 * line_covered / line_total if line_total else 0.0,
        "branch": 100.0 * branch_covered / branch_total if branch_total else 0.0,
    }


def compare_againsts_baseline(current: Dict[str, float], baseline: Dict[str, float]) -> Dict[str, object]:
    tolerance = 0.3
    raise_for_improvement = 1.0
    epsilon = 1e-9

    for metric in ("line", "branch"):
        delta = current[metric] - baseline[metric]
        if delta < -(tolerance + epsilon):
            return {
                "status": "fail",
                "message": f"Coverage regression detected for {metric}: current {current[metric]:.1f}% is {baseline[metric] - current[metric]:.1f} percentage points below baseline {baseline[metric]:.1f}%.",
            }
        if delta > raise_for_improvement + epsilon:
            return {
                "status": "fail",
                "message": f"Coverage improved for {metric}: current {current[metric]:.1f}% exceeds baseline {baseline[metric]:.1f}% by {delta:.1f} points. Please update backend/coverage-baseline.json in this PR.",
            }

    return {"status": "pass", "message": "Coverage is within the ratchet bounds."}


def main() -> int:
    if len(sys.argv) != 3:
        print("Usage: python scripts/check_coverage_ratchet.py <jacoco.csv> <coverage-baseline.json>", file=sys.stderr)
        return 2

    csv_path = Path(sys.argv[1]).resolve()
    baseline_path = Path(sys.argv[2]).resolve()

    if not csv_path.exists():
        print(f"JaCoCo CSV report not found: {csv_path}", file=sys.stderr)
        return 2

    if not baseline_path.exists():
        print(f"Coverage baseline file not found: {baseline_path}", file=sys.stderr)
        return 2

    current = load_current_coverage(csv_path)
    baseline = load_baseline(baseline_path)
    result = compare_againsts_baseline(current, baseline)
    print(json.dumps({"current": current, "baseline": baseline, "result": result}, indent=2))
    return 0 if result["status"] == "pass" else 1


if __name__ == "__main__":
    sys.exit(main())

File: frontend/scripts/check-coverage-ratchet.mjs
#!/usr/bin/env node

import { readFile } from 'node:fs/promises';
import process from 'node:process';

const [, , summaryPath = 'coverage/coverage-summary.json', baselinePath = 'coverage-baseline.json'] = process.argv;

const TOLERANCE_PCT = 0.3;
const RAISE_FOR_IMPROVEMENT_PCT = 1.0;

// Suelos absolutos recomendados para proyectos React/TypeScript con Testing Library.
// El baseline dinamico tambien debe mantenerse siempre por encima de estos valores.
const MINIMUM_LINE_PCT = 70;
const MINIMUM_BRANCH_PCT = 55;

const readJson = async (path) => JSON.parse(await readFile(path, 'utf8'));

const roundOne = (value) => Math.round((value + Number.EPSILON) * 10) / 10;

const readCurrentCoverage = (summary) => ({
    line: Number(summary?.total?.lines?.pct),
    branch: Number(summary?.total?.branches?.pct),
});

const validateCoverage = (coverage, label) => {
    for (const metric of ['line', 'branch']) {
        if (!Number.isFinite(coverage[metric])) {
            throw new Error(`${label} coverage is missing a numeric ${metric} value.`);
        }
    }
};

const compareWithBaseline = (current, baseline) => {
    const currentRounded = {
        line: roundOne(current.line),
        branch: roundOne(current.branch),
    };
    const baselineRounded = {
        line: roundOne(baseline.line),
        branch: roundOne(baseline.branch),
    };

    for (const metric of ['line', 'branch']) {
        const delta = currentRounded[metric] - baselineRounded[metric];
        if (delta < -TOLERANCE_PCT) {
            return {
                status: 'fail',
                message: `Coverage regression detected for ${metric}: current ${currentRounded[metric].toFixed(1)}% is ${(baselineRounded[metric] - currentRounded[metric]).toFixed(1)} percentage points below baseline ${baselineRounded[metric].toFixed(1)}%.`,
            };
        }
        if (delta > RAISE_FOR_IMPROVEMENT_PCT) {
            return {
                status: 'fail',
                message: `Coverage improved for ${metric}: current ${currentRounded[metric].toFixed(1)}% exceeds baseline ${baselineRounded[metric].toFixed(1)}% by ${delta.toFixed(1)} points. Please update ${baselinePath} in this PR.`,
            };
        }
    }

    return {
        status: 'pass',
        message: 'Coverage is within the ratchet bounds.',
    };
};

const checkMinimumThresholds = (current) => {
    const failures = [];

    if (current.line < MINIMUM_LINE_PCT) {
        failures.push(`Line coverage ${current.line.toFixed(1)}% is below the minimum required threshold of ${MINIMUM_LINE_PCT}%.`);
    }
    if (current.branch < MINIMUM_BRANCH_PCT) {
        failures.push(`Branch coverage ${current.branch.toFixed(1)}% is below the minimum required threshold of ${MINIMUM_BRANCH_PCT}%.`);
    }

    return failures.length === 0
        ? { status: 'pass', message: 'Coverage meets the absolute minimum thresholds.' }
        : { status: 'fail', message: failures.join(' ') };
};

const main = async () => {
    const summary = await readJson(summaryPath);
    const baseline = await readJson(baselinePath);
    const current = readCurrentCoverage(summary);

    validateCoverage(current, 'Current');
    validateCoverage(baseline, 'Baseline');

    // Las dos validaciones se ejecutan siempre y se informan por separado.
    const ratchet = compareWithBaseline(current, baseline);
    const thresholds = checkMinimumThresholds(current);
    const overall = ratchet.status === 'pass' && thresholds.status === 'pass' ? 'pass' : 'fail';

    console.log(JSON.stringify({
        current,
        baseline,
        minimums: {
            line: MINIMUM_LINE_PCT,
            branch: MINIMUM_BRANCH_PCT,
        },
        result: {
            ratchet,
            thresholds,
            overall,
        },
    }, null, 2));

    process.exitCode = overall === 'pass' ? 0 : 1;
};

main().catch((error) => {
    console.error(error instanceof Error ? error.message : String(error));
    process.exitCode = 2;
});
